lib/dRep/dRepImpl.py

```diff
@@ -0,0 +1,441 @@
+# -*- coding: utf-8 -*-
+#BEGIN_HEADER
+import logging
+import os
+import sys
+import shutil
+import subprocess
+import pprint
+import uuid
+import re
+import functools
+
+from installed_clients.KBaseReportClient import KBaseReport
+from installed_clients.DataFileUtilClient import DataFileUtil
+from installed_clients.WorkspaceClient import Workspace
+from installed_clients.MetagenomeUtilsClient import MetagenomeUtils
+
+from .util import report
+from .util.dprint import dprint
+from .util.kbase_obj import BinnedContigs
+from .util.config import _globals, reset
+
+#END_HEADER
+
+
+class dRep:
+    '''
+    Module Name:
+    dRep
+
+    Module Description:
+    A KBase module: dRep
+    '''
+
+    ######## WARNING FOR GEVENT USERS ####### noqa
+    # Since asynchronous IO can lead to methods - even the same method -
+    # interrupting each other, you must be *very* careful when using global
+    # state. A method could easily clobber the state set by another while
+    # the latter method is running.
+    ######################################### noqa
+    VERSION = "0.0.1"
+    GIT_URL = "https://github.com/n1mus/dRep.git"
+    GIT_COMMIT_HASH = "05f421114e50a04008d9897cdb20399b23a47b46"
+
+    #BEGIN_CLASS_HEADER
+    #END_CLASS_HEADER
+
+    # config contains contents of config file in a hash or None if it couldn't
+    # be found
+    def __init__(self, config):
+        #BEGIN_CONSTRUCTOR
+
+        callback_url = os.environ['SDK_CALLBACK_URL']
+        workspace_url = config['workspace-url']
+        shared_folder = config['scratch']
+        
+        self._globals = { # shared by all API-method runs
+            'shared_folder': config['scratch'], 
+            'ws': Workspace(workspace_url),
+            'dfu': DataFileUtil(callback_url),
+            'mgu': MetagenomeUtils(callback_url),
+            'kbr': KBaseReport(callback_url),
+            }
+
+        logging.basicConfig(format='%(created)s %(levelname)s: %(message)s',
+                            level=logging.INFO)
+        #END_CONSTRUCTOR
+        pass
+
+
+    def dereplicate(self, ctx, params):
```

Comment:
This function is very very long and it is easy to lose track of what context we are in. I would shorten the function as much as possible and move discrete functionality into utility modules and functions

Reply:
hmm it's 344 lines and usually is only 1 or 2 indents in. i thought the triangular configuration of hashes created visual logic blocks. it's striking on gvim

i guess it would be a good idea if there were more API-methods added